### ISSUE TITLE
uniqMerge function in connector

### ISSUE THREAD

**Issue by vlauzeckas** [analytikaplus/clickhouse-tableau-connector-jdbc#47]
hi,
Clickhouse have unique and very powerful functionality called Aggregate state. Basically, you can aggregate "count distinct". More about this functionality with examples could found here:
https://altinity.com/blog/2017/7/10/clickhouse-aggregatefunctions-and-aggregatestate

But due to Tableau limitations its not possible to use it properly. Its possible to be used by using RAWSQL functions. Example:
`RAWSQLAGG_INT("uniqMerge(%1)" , [MyKPI])`

Problem is that  "RAWSQL pass-through functions will not work with published data sources or with Tableau extracts.":
https://help.tableau.com/current/pro/desktop/en-us/functions_functions_passthrough.htm

I wonder if it would be possible to create such a function on the connector level? Like you did with "RAND()", "UNHEX([my_string])" and etc. 

This aggregatestate are very powerful functionality of clickhouse, it is very pity you can't create data sources using it.

**Comment by artshevchenko** [comment]
Hi! Of course we can add this function to the dialect. I would be grateful for a complete list of all the necessary functions with the -Merge combinator, so it will be easier to implement this functionality. As far as I understand, the implementation of functions with the -State combinator is not required, is it?

**Comment by vlauzeckas** [comment]
very good to hear this!
its hard for me to foresee what others may need, but i think those functions should cover needs of most of the cases:
```
count
min
max
sum
avg
uniq
uniqExact
```

And yes, only with merge combinators.
so we will have uniqMerge, minMerge, sumMege and etc.

**Comment by vlauzeckas** [comment]
@artshevchenko we badly need this Aggregate state functionality, I am just not 100% sure that creating those functions will solve the problem. It works perfectly fine on Tableau desktop, but stops working once you publish the data source to the server. We are ready to pay to get this functionality, maybe it is possible to hire you to help us solve this? Is there a way to contact you?

**Comment by artshevchenko** [comment]
@vlauzeckas Hi! At the moment, the problem is not in financing, we have a technical problem. The AggregateFunction data type is not supported by the JDBC driver and this breaks the View Data functionality. When adding a table with an AggregateFunction column to the Data Source, this column is displayed as a String in Tableau, and the only way to use this column is to create a new Calculated Field with an expression like COUNTD_MERGE([AggregateFunction]). Any other scenarios lead to errors. If adding new features will be enough for you, we can soon send a beta version of the connector with this functionality

**Comment by vlauzeckas** [comment]
Yes, i understand that this should be used very explicitly.

**Comment by juliusstep** [comment]
Hi @artshevchenko, 

As @vlauzeckas noted, we understand the limitations of adding uniqMerge() functionality. And understand that implementation of such function might break expected behavior in Tableau Desktop when previewing data. 
We could try to adjust tdd file and build taco file by our self. But we can't sign that file. And to use unsigned connector in Tableau Desktop and Server we would need to overwrite settings by disabling plugin signature verification. Which is currently not possible on our side.
So we are more than ok if such functions would be added as experimental functionality in upcoming release with a notes in README.

**Comment by artshevchenko** [comment]
@vlauzeckas @juliusstep Hi! Please share your feedback on this beta version of the connector. In this version, the functions `COUNTD_MERGE()` and `COUNTD_UNIQ_MERGE()` have been added.
[clickhouse_jdbc_0.2.6.1.taco.zip](https://github.com/analytikaplus/clickhouse-tableau-connector-jdbc/files/11491028/clickhouse_jdbc_0.2.6.1.taco.zip)

**Comment by vlauzeckas** [comment]
We tested it out and it works perfectly. if you could release this, we could update connector and reports on our production servers

**Comment by vlauzeckas** [comment]
is there any plan to release new version with those changes?

**Comment by hipoagu** [comment]
I am having problems to use AggregateFunction fields when the datasource is not published, but straight live into Tableau Desktop.
If the table with the aggregated function fields is dropped as a whole into the datasource editor it works fine, but if it is converted to Custom SQL (and i need to do so to apply some custom selection of columns, renames etc) then Tableau Throws exception.
With Altinity ODBC connector/driver it works perfectly either whole table dropped or with custom SQL to that very same table.

**Comment by yurifal** [comment]
What if you create a view (from the Custom SQL)?
Or may be wrap the Custom SQL with the outer SELECT? 
`SELECT * FROM ( your_Custom_SQL )`

**Comment by yurifal** [comment]
```
CREATE VIEW test_aggfunction AS (
SELECT 
  "mv"."reporting_date",
  "mv"."avg_2g_rxlev" AS "avg_2g_rxlev"
FROM "cellrebel"."mv_rf_dimensions_1km_daily_mea" AS "mv"
GROUP BY "mv"."reporting_date"
) ;
```

then use this view as a datasource in Tableau.

to merge from the State column ( avg_2g_rxlev in your example ) 
use the following RAWSQL() calculated field in Tableau
```
RAWSQL_REAL("
avgMerge(%1),
[avg_2g_rxlev]
)  
```

**Comment by vlauzeckas** [comment]
hi Yuri,
thanks for working on this. If i recall correctly you can't publish data
source to tableau server with "pass through" sql functions. that was
issue with this approach.

V.

**Comment by yurifal** [comment]
> hi Yuri,
> thanks for working on this. If i recall correctly you can't publish data
> source to tableau server with "pass through" sql functions. that was
> issue with this approach.
> 
> V.
> […](#)

Yep, exactly this. That's why the `COUNTD_MERGE()` and `COUNTD_UNIQ_MERGE()` functions have been added to this fork of the connector. 

The example above is using the `AVG` aggregate (I guess), hence the RAWSQL (just for testing purposes).

**Comment by hipoagu** [comment]
Hello,

Tried the view approach, it did the trick! However the group by had to be removed as clickhouse would complain about avg_2g_rxlev not being part of it. 

Just this worked:

CREATE VIEW test_aggfunction AS (
SELECT 
  "mv"."reporting_date",
  "mv"."avg_2g_rxlev" AS "avg_2g_rxlev"
FROM "cellrebel"."mv_rf_dimensions_1km_daily_mea" AS "mv"
) ;

Then from Tableau i dropped the view to my datasource and omit the Customization of SQL.

Thank you very much. Does this approach then also allow to publish the datasource and use it as published datasource from other workbooks?

However still dissapointed that Altinity's ODBC connector does allow this custom SQL from Tableau while official Clickhouse JDBC does not!

**Comment by hipoagu** [comment]
> Hello,
> 
> Tried the view approach, it did the trick! However the group by had to be removed as clickhouse would complain about avg_2g_rxlev not being part of it.
> 
> Just this worked:
> 
> CREATE VIEW test_aggfunction AS ( SELECT "mv"."reporting_date", "mv"."avg_2g_rxlev" AS "avg_2g_rxlev" FROM "cellrebel"."mv_rf_dimensions_1km_daily_mea" AS "mv" ) ;
> 
> Then from Tableau i dropped the view to my datasource and omit the Customization of SQL.
> 
> Thank you very much. Does this approach then also allow to publish the datasource and use it as published datasource from other workbooks?
> 
> However still dissapointed that Altinity's ODBC connector does allow this custom SQL from Tableau while official Clickhouse JDBC does not!

I just answer myself wrt the published datasource: No, it doesn't work if datasource is published. I need to keep the datasource embeded in workbook as live connection to the view that wraps the materialized view that contains aggregatefunction fields.

**Comment by yurifal** [comment]
I guess the "cellrebel"."mv_rf_dimensions_1km_daily_mea" matview 
to be built using the AggregatingMergeTree engine. 
If yes, no need for the GROUP BY.
